src/intervals/intervals.jl

```diff
@@ -144,6 +144,14 @@ function ..(a::T, b::S) where {T, S}
     interval(atomic(Interval{T}, a).lo, atomic(Interval{S}, b).hi)
 end
 
+function ..(a::T, b::S) where {T<:Integer, S<:AbstractFloat}
+    interval(atomic(Interval{S}, a).lo, atomic(Interval{S}, b).hi)
+end
+
+function ..(a::T, b::S) where {T<:AbstractFloat, S<:Integer}
+    interval(atomic(Interval{T}, a).lo, atomic(Interval{T}, b).hi)
```

Comment:
Can't you just use promotion somehow?

Reply:
I could have use `promote_type` but in my understanding having more methods is more efficient than relying on functions that deal with types.
```Julia
function ..(a::T, b::S) where {T, S}
    R = promote_type(T, S)
    interval(atomic(Interval{R}, a).lo, atomic(Interval{R}, b).hi)
end
```
should work. Yet it probably is inefficient for `..(::Floatxx, ::Floatyy)`.

Reply:
I'm not sure what you mean by "having more methods is more efficient than relying on functions that deal with types".  Julia compiles functions to make specialised versions corresponding to the types of the inputs.

I'm not sure why you think that this code will be inefficient.
Rather than guessing what might be inefficient, just use BenchmarkTools.jl to test the different possibilities to see which actually are faster or slower.

Reply:
I though that calling `promote_type` would have been slow. But, you were write:
```Julia
julia> using BenchmarkTools

julia> using IntervalArithmetic

julia> function foo(a::T, b::S) where {T, S}
           interval(IntervalArithmetic.atomic(Interval{T}, a).lo, IntervalArithmetic.atomic(Interval{S}, b).hi)
       end
foo (generic function with 1 method)

julia> function foo(a::T, b::S) where {T<:Integer, S<:AbstractFloat}
           interval(IntervalArithmetic.atomic(Interval{S}, a).lo, IntervalArithmetic.atomic(Interval{S}, b).hi)
       end
foo (generic function with 2 methods)

julia> function foo(a::T, b::S) where {T<:AbstractFloat, S<:Integer}
           interval(IntervalArithmetic.atomic(Interval{T}, a).lo, IntervalArithmetic.atomic(Interval{T}, b).hi)
       end
foo (generic function with 3 methods)

julia> function foo(a::T, b::Irrational{S}) where {T, S}
           R = promote_type(T, Irrational{S})
           interval(IntervalArithmetic.atomic(Interval{R}, a).lo, R(b, RoundUp))
       end
foo (generic function with 4 methods)

julia> function foo(a::Irrational{T}, b::S) where {T, S}
           R = promote_type(Irrational{T}, S)
           return Interval(R(a, RoundDown), IntervalArithmetic.atomic(Interval{R}, b).hi)
       end
foo (generic function with 5 methods)

julia> function foo(a::Irrational{T}, b::Irrational{S}) where {T, S}
           R = promote_type(Irrational{T}, Irrational{S})
           return Interval(a, b)
       end
foo (generic function with 6 methods)

julia> function bar(a::T, b::S) where {T, S}
           R = promote_type(T, S)
           interval(IntervalArithmetic.atomic(Interval{R}, a).lo, IntervalArithmetic.atomic(Interval{R}, b).hi)
       end
bar (generic function with 1 method)

julia> function bar(a::T, b::Irrational{S}) where {T, S}
           R = promote_type(T, Irrational{S})
           interval(IntervalArithmetic.atomic(Interval{R}, a).lo, R(b, RoundUp))
       end
bar (generic function with 2 methods)

julia> function bar(a::Irrational{T}, b::S) where {T, S}
           R = promote_type(Irrational{T}, S)
           return Interval(R(a, RoundDown), IntervalArithmetic.atomic(Interval{R}, b).hi)
       end
bar (generic function with 3 methods)

julia> function bar(a::Irrational{T}, b::Irrational{S}) where {T, S}
           R = promote_type(Irrational{T}, Irrational{S})
           return Interval(a, b)
       end
bar (generic function with 4 methods)

julia> @benchmark foo(2, 3)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.207 ns (0.00% GC)
  median time:      2.219 ns (0.00% GC)
  mean time:        2.307 ns (0.00% GC)
  maximum time:     21.396 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000

julia> @benchmark bar(2, 3)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.208 ns (0.00% GC)
  median time:      2.221 ns (0.00% GC)
  mean time:        2.273 ns (0.00% GC)
  maximum time:     26.497 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000

julia> @benchmark foo(2.5, 3)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     212.750 ns (0.00% GC)
  median time:      218.305 ns (0.00% GC)
  mean time:        223.686 ns (0.00% GC)
  maximum time:     434.141 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     525

julia> @benchmark bar(2.5, 3)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     216.353 ns (0.00% GC)
  median time:      219.258 ns (0.00% GC)
  mean time:        227.008 ns (0.00% GC)
  maximum time:     503.979 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     516
```